Mage.Sets/src/mage/cards/s/SmugglersShare.java

```diff
@@ -0,0 +1,131 @@
+import mage.constants.TargetController;
+import mage.game.Game;
+import mage.game.permanent.Permanent;
+import mage.game.permanent.token.TreasureToken;
+import mage.players.Player;
+import mage.watchers.common.CardsAmountDrawnThisTurnWatcher;
+import mage.watchers.common.PermanentsEnteredBattlefieldWatcher;
+
+/**
+ * One way or another, everyone gets paid what they're owed.
+ * Based on Gadrak, the Crown Scourge, Hoard Hauler, Lavaball Trap, Walker Of The Grove, Widespread Thieving, and Wound Reflection
+ *
+ * @author Slanman3755
+ */
+public final class SmugglersShare extends CardImpl {
+
+    public SmugglersShare(UUID ownerId, CardSetInfo setInfo) {
+        super(ownerId, setInfo, new CardType[]{CardType.ENCHANTMENT}, "{2}{W}");
+
+        // At the beginning of each end step, draw a card for each opponent who drew two or more cards this turn, then
+        // create a Treasure token for each opponent who had two or more lands enter the battlefield under their control
+        // this turn.
+        Ability ability = new BeginningOfEndStepTriggeredAbility(new DrawCardSourceControllerEffect(SmugglersShareDrawValue.instance), TargetController.EACH_PLAYER, false);
+        ability.addEffect((new CreateTokenEffect(new TreasureToken(), SmugglersShareTreasureValue.instance)).concatBy(", then"));
+        ability.addWatcher(new CardsAmountDrawnThisTurnWatcher());
+        ability.addWatcher(new PermanentsEnteredBattlefieldWatcher());
+        this.addAbility(ability);
+    }
+
+    private SmugglersShare(final SmugglersShare card) {
+        super(card);
+    }
+
+    @Override
+    public SmugglersShare copy() {
+        return new SmugglersShare(this);
+    }
+}
+
+enum SmugglersShareDrawValue implements DynamicValue {
+    instance;
+
+    @Override
+    public int calculate(Game game, Ability source, Effect effect) {
+        Player controller = game.getPlayer(source.getControllerId());
+        CardsAmountDrawnThisTurnWatcher watcher = game.getState().getWatcher(CardsAmountDrawnThisTurnWatcher.class);
+        if (watcher == null) {
+            return 0;
+        }
+        int drawCount = 0;
+        for (UUID playerId : game.getOpponents(controller.getId())) {
+            Player opponent = game.getPlayer(playerId);
+            if (opponent != null && watcher.getAmountCardsDrawn(playerId) >= 2) {
+                drawCount++;
+            }
+        }
+        return drawCount;
+    }
+
+    @Override
+    public SmugglersShareDrawValue copy() {
+        return instance;
+    }
+
+    @Override
+    public String getMessage() {
+        return "opponent who drew two or more cards this turn";
+    }
+
+    @Override
+    public String toString() {
+        return "1";
+    }
+}
+
+enum SmugglersShareTreasureValue implements DynamicValue {
+    instance;
+
+    @Override
+    public int calculate(Game game, Ability source, Effect effect) {
+        Player controller = game.getPlayer(source.getControllerId());
```

Comment:
Same as above. Need to check for null

Reply:
Can game.getPlayer return null or source.getControllerId return null? or both?

Reply:
Source itself should never be null.
game.getPlayer is fine will null input, it checks for it and returns null.
game.getPlayer is the issue, it will sometimes return null.

Reply:
Some implementations just call source.getControllerId() directly instead of getPlayer(source.getControllerId()).getId() or whatever. Should we do that instead?

Reply:
You can grab the player to check if they're null. And shortcut the whole thing.

Otherwise I think calling source.getController when making the for loop will also work since there will be nothing to iterate over and it will return 0.

I would personally do the check, that way I can code the rest of the function knowing that player exists. Potentially prevents a mistake in returning the wrong value (not an issue in this case)

Reply:
I will leave it as is then with the check in place.